### PR TITLE
lazy_import as a context manager

### PR DESCRIPTION
**Pull request by maurosilber**
Hi!

I tried a different approach to solve this problem,
which defers a standard import statement inside a context manager.

Usage:

```python
from lazy_loader import lazy_import

with lazy_import():
    import math  # lazy

math.pi  # executes module
```

Advantages:
- possibly simpler to adopt and explain,
- it eagerly raises an `ImportError` if a module is not found,
- it works with static type checkers, providing type hints and completions.

I'm not sure if it is missing some functionality that the current version of `lazy_loader` has.

How it works:

It injects LazyFinder to sys.meta_path, which searches with the rest of sys.meta_path's Finders.
If it found one, it wraps the resulting ModuleSpec in importlib.util.LazyLoader, which defers loading until first access.

I hope you can check it out!

### CODE REVIEW

**Comment by maurosilber** [comment]
I had never heard of `sys.meta_path` either until today, but here it is on the docs: https://docs.python.org/3/library/sys.html#sys.meta_path

According to [PEP 451](https://peps.python.org/pep-0451), which introduced ModuleSpecs in Python 3.4, it seems `sys.meta_path` was introduced with [PEP 302](https://peps.python.org/pep-0302) for Python 2.3 in 2002.

**Comment by maurosilber** [comment]
> Can you please make a PR that doesn't take out all the old code, but adds this as a new function? That'd make the diff easier to review.

Done!

**Comment by tlambert03** [comment]
I like the syntax too!  but are you sure it's lazy?

I just checked out this PR, installed it, and created the following test case:

```
mod/
├── __init__.py
├── __init__.pyi
└── _mod.py
```

where `_mod.py` contains
```python
print("IMPORT")

def some_func(x: int) -> int:
    return x + 1
```

### `lazy_loader.attach`

```python
# mod/__init__.py
import lazy_loader

__getattr__, __dir__, __all__ = lazy_loader.attach(
    __name__, submod_attrs={"_mod": ["some_func"]}
)
```

```python
In [1]: import mod

In [2]: dir(mod)
Out[2]: ['some_func']

In [3]: mod.some_func
IMPORT
Out[3]: <function some_func at 0x1038af130>
```

### `lazy_loader.attach_stub`

```python
# mod/__init__.py
import lazy_loader

__getattr__, __dir__, __all__ = lazy_loader.attach_stub(__name__, __file__)
```

```python
In [1]: import mod

In [2]: dir(mod)
Out[2]: ['some_func']

In [3]: mod.some_func
IMPORT
Out[3]: <function some_func at 0x107fbbeb0>
```

### `lazy_loader.lazy_import`

```python
# mod/__init__.py
import lazy_loader

with lazy_loader.lazy_import():
    from ._mod import some_func
```

‼️ greedy import
```python
In [1]: import mod
IMPORT

In [2]: dir(mod)
Out[2]:
[
    '__builtins__',
    '__cached__',
    '__doc__',
    '__file__',
    '__loader__',
    '__name__',
    '__package__',
    '__path__',
    '__spec__',
    '_mod',
    'lazy_loader',
    'some_func'
]
```

**Comment by tlambert03** [comment]
I do think that adding a `MetaPathFinder` to `sys.path` would potentially be another potentially elegant way to solve this, but i'm not sure this is doing what we want it to?

**Comment by tlambert03** [comment]
> it eagerly raises an ImportError if a module is not found,

I think this is because it's actually eagerly importing?

**Comment by maurosilber** [comment]
But `from ._mod import some_func` is accesing an attribute of `._mod`, hence it imports `_mod`.

You need to do:

```python
import lazy_loader

with lazy_loader.lazy_import():
    import mod  # lazy

mod.some_func  # triggers import
```

as in the examples where you used `lazy_loader.attach`.

**Comment by maurosilber** [comment]
> > it eagerly raises an ImportError if a module is not found,
> 
> I think this is because it's actually eagerly importing?

The import system seems to be divided in finding and loading. As it doesn't find the module, it can raise an eager `ImportError`. But if it is found, it isn't yet loaded, waiting until first attribute access..

**Comment by tlambert03** [comment]
I'm a bit confused... in the current implement of lazy.attach, the goal is to be able to do something like:

```
>>> import mod   # mod._mod not yet imported
>>> mod.some_func()  # on attribute access, NOW mod._mod is imported and some_func taken
```

are you suggesting that it would be `mod.mod.some_func`?    and doesn't [changing the test here](https://github.com/scientific-python/lazy_loader/pull/11/files#diff-fa2a9225e8fc9a8bb42d66a374408bc3c3643236092b53e71993fe71b3c8cc2eR95) to `assert isinstance(fake_pkg.some_func, types.ModuleType)` suggest that you've changed what [SPEC 1](https://scientific-python.org/specs/spec-0001/) is trying to achieve?

**Comment by maurosilber** [comment]
Ok, now I understand that that test.

Yes, you're right. What I did is different to SPEC 1, and only works for modules, but not for exported functions.

For instance, if `skimage.__init__.py` were:
```python
from lazy_loader import lazy_import

with lazy_import():
    from . import filters, morphology, ...
```
it would lazily export all those subpackages. And then, you could do:
```python
import skimage as ski  # cheap; does not load submodules

skimage.filters  # only loads the filters submodule
```
where `skimage.filters` is lazily loaded, and the same is valid for any submodule inside `skimage.filters` (which is also inside a `with lazy_import()`, of course).

But [`skimage.filters.__init__.py`](https://github.com/scikit-image/scikit-image/blob/main/skimage/filters/__init__.py) only exports a single module, `rank` and then exports all functions, which are equivalent to `from .submodule import function` and would be eagerly loaded by what I propose.

So, in conclusion, it does not replace `attach`. But it could take over the `submodules` parameter of `attach`, wouldn't it?

**Comment by tlambert03** [comment]
> So, in conclusion, it does not replace attach. But it could take over the submodules parameter of attach, wouldn't it?

yep, that jives with my understanding of what it's doing
